File: sync.js
'use strict';
const fs = require('fs');
const { join, basename } = require('path');
const {
  convertToRelative,
  mkdir,
  stat,
  copyFile,
  pathExists
} = require('./utils/file-util');
const readlineSync = require('readline-sync');
const { deleteDeep, walkdir } = require('more-node-fs');
const { ArgumentParser } = require('argparse');

// Get command line arguments:
const argparser = new ArgumentParser({ description: 'Sync-Local-and-Cloud' });
argparser.add_argument('config', {
  help: 'The path to the config JSON file.'
});
argparser.add_argument('manifest', {
  help: 'The path to the manifest JSON file.'
});
argparser.add_argument('-t', '--test', {
  default: false, action: 'store_true'
});
argparser.add_argument('-s', '--select', {
  default: false,
  type: arg => arg.toLowerCase().trim() === 'true'
});
argparser.add_argument('-a', '--action', {
  default: 'sync',
  type: arg => {
    arg = arg.toLowerCase().trim();
    if (!['sync', 'pull', 'push'].includes(arg)) {
      throw new Error('action parameter is invalid.');
    }
    return arg;
  }
});
const {
  config: configPath, manifest: manifestPath, select, action, test
} = argparser.parse_args();

const config = require(configPath);
const manifestFile = require(manifestPath);
const CONFIG_RUN_SELECTION = [];
if (test) {
  process.env.TEST = true;
  console.log(' * Running in TEST mode');
}
if (select) {
  let input = 0;
  while (input > 0 || !CONFIG_RUN_SELECTION.length) {
    console.clear();
    console.log('Selected:', CONFIG_RUN_SELECTION);
    const selection = [
      'RUN',
      ...config
        .filter(({ active, repoName }) => {
          return active && !CONFIG_RUN_SELECTION.includes(repoName);
        })
        .map(({ repoName }) => repoName)
    ];
    input = readlineSync.keyInSelect(selection, 'Which config to run?');
    if (input === -1) {
      console.log('Aborting...');
      return;
    } else if (input > 0) {
      CONFIG_RUN_SELECTION.push(selection[input]);
    }
  }
}

async function directoryFunc (dir, root, otherRoot, manifest) {
  if (dir === root)
    return;
  const otherDir = join(otherRoot, convertToRelative(dir, root));

  const dirIsDeleted = manifest[root] && // A check to see if initialising for the first time
    ((manifest[dir] === false && !await pathExists(dir, 0)) ||
    (manifest[otherDir] === false && !await pathExists(otherDir, 0)));

  if (dirIsDeleted) {
    await deleteDeep(dir);
    await deleteDeep(otherDir);
    manifest[dir] = manifest[otherDir] = false;
  // If the directory at path doesn't exists then make it:
  } else if (!await pathExists(otherDir, 0)) {
    await mkdir(otherDir, manifest);
    manifest[dir] = true;
  }
}

async function fileFunc (file, root, otherRoot, manifest) {
  const otherFile = join(otherRoot, convertToRelative(file, root));

  const fileIsDeleted = manifest[root] && // A check to see if initialising
    ((manifest[file] === false && !await pathExists(file, 0)) ||
    (manifest[otherFile] === false && !await pathExists(otherFile, 0)));

  if (fileIsDeleted) {
    await deleteDeep(file);
    await deleteDeep(otherFile);
    manifest[file] = manifest[otherFile] === false;
  } else if (!await pathExists(otherFile, 0))
    await copyFile(file, otherFile, manifest);
  else {
    const stats = await stat(file);
    const otherStats = await stat(otherFile);
    // Compare stat modified times:
    if (stats.mtimeMs > otherStats.mtimeMs)
      await copyFile(file, otherFile, manifest);
    else if (stats.mtimeMs < otherStats.mtimeMs)
      await copyFile(otherFile, file, manifest);
  }
}

/**
 * @param {string} str
 * @returns {RegExp} Returns the regexString parameter parsed as an instance of RegExp.
 */
function parseRegExp (str) {
  if (!str)
    return null;
  const flags = str.substring(str.lastIndexOf('/') + 1);
  const source = str.substring(1, str.lastIndexOf('/'));
  return new RegExp(source, flags);
}

async function updateManifest (manifest, ignore, matchFile) {
  for (const path in manifest) {
    if (
      manifest[path] === false ||
      (ignore instanceof RegExp && ignore.test(manifest[path])) ||
      (matchFile instanceof RegExp && !matchFile.test((basename(manifest[path]))))
    )
      delete manifest[path]; // Remove old deleted flagged paths or if it's being ignored
    else if (!await pathExists(path, 0))
      manifest[path] = false; // Flag as deleted.
  }
}

function readdir (path, ignore, matchFile) {
  const result = { files: [], dirs: [] };
  for (const { path: p, stats } of walkdir(path, { ignore })) {
    if (stats.isDirectory()) {
      result.dirs.push(p);
    } else if (!matchFile || (matchFile instanceof RegExp && matchFile.test(basename(p)))) {
      result.files.push(p);
    }
  }
  return result;
}

(async () => {
  console.time('sync.js');

  await Promise.all(config.map(async ({
    cloudDirPath, localDirPath, repoName, ignore, matchFile, active
  }) => {
    if (!active || (CONFIG_RUN_SELECTION.length && !CONFIG_RUN_SELECTION.includes(repoName)))
      return;
    console.log(`${action}ing ${repoName}`);
    console.time(repoName);

    let manifest;
    if (!manifestFile[repoName])
      manifest = manifestFile[repoName] = {};
    else
      manifest = manifestFile[repoName];

    ignore = parseRegExp(ignore);
    matchFile = parseRegExp(matchFile);

    cloudDirPath = cloudDirPath.replace(/\//g, '\\');
    localDirPath = localDirPath.replace(/\//g, '\\');
    if (!await pathExists(cloudDirPath, 0))
      await mkdir(cloudDirPath, manifest);
    if (!await pathExists(localDirPath, 0))
      await mkdir(localDirPath, manifest);

    const cloud = readdir(cloudDirPath, ignore, matchFile);
    const local = readdir(localDirPath, ignore, matchFile);

    await updateManifest(manifest, ignore, matchFile);

    if (['sync', 'pull'].includes(action)) {
      for (const dir of cloud.dirs)
        await directoryFunc(dir, cloudDirPath, localDirPath, manifest);
    }
    if (['sync', 'push'].includes(action)) {
      for (const dir of local.dirs)
        await directoryFunc(dir, localDirPath, cloudDirPath, manifest);
    }

    if (['sync', 'pull'].includes(action)) {
      for (const file of cloud.files)
        await fileFunc(file, cloudDirPath, localDirPath, manifest);
    }
    if (['sync', 'push'].includes(action)) {
      for (const file of local.files)
        await fileFunc(file, localDirPath, cloudDirPath, manifest);
    }

    await updateManifest(manifest, ignore, matchFile);
    manifest[cloudDirPath] = manifest[localDirPath] = true;

    console.timeEnd(repoName);
  }));

  // Overwrite manifest file:
  if (!process.env.TEST) {
    fs.writeFileSync('./manifest.json', JSON.stringify(manifestFile, null, 2));
    console.log('Saved manifest.json');
  }

  console.timeEnd('sync.js');
})();

File: sync.nim
import os, strformat, times, utils/fileutils, re, cligen, json, tables, strutils, threadpool
{.experimental: "parallel".}

var test: bool

proc parseRegex (regex: string): Regex =
  if regex == "":
    return re"(?=a)b"
  if not match(regex, re"^/.+/.*$"):
    raise newException(IOError, $fmt("Regex parse error on string: {regex}"))
  
  let lastSlash = regex.rfind('/') - 1
  var flags: set[RegexFlag]
  for c in regex[lastSlash + 1 .. regex.high]:
    if c == 'i':
      flags = flags + {reIgnoreCase}
    # Not many other nim regex flags to support
  return re(regex[1 .. lastSlash], flags)

proc updateManifest (manifest: JsonNode, ignore: Regex) =
  for path, exists in manifest.getFields().pairs:
    if not getBool(exists) or contains(path, ignore):
      manifest.delete(path) # Remove old deleted flagged paths or if it's being ignored
    elif existsPath(path) == 0:
      manifest[path] = newJBool false # Flag as deleted.

proc directoryFunc (
  dir: string, root: string, otherRoot: string, manifest: JsonNode
) =
  if dir == root:
    return
  let otherDir = otherRoot / relativePath(dir, root)
  let dirExists = existsDir dir
  let otherDirExists = existsDir otherDir

  let dirIsDeleted = manifest{root}.getBool() and # A check to see if initialising for the first time
    ((manifest{dir} != nil and manifest{dir}.getBool() == false and not dirExists) or
    (manifest{otherdir} != nil and manifest{otherDir}.getBool() == false and not otherDirExists))

  if dirIsDeleted:
    if dirExists:
      echo fmt" - '{dir}'"
    if otherDirExists:
      echo fmt" - '{otherDir}'"
    if not test:
      removeDir dir
      removeDir otherDir
    manifest{dir} = newJBool false
    manifest{otherDir} = newJBool false
  elif not existsDir otherDir:
    # If the directory at otherDir doesn't exist then make it:
    echo fmt" + '{otherDir}'"
    if not test:
      createDir otherDir
    manifest{dir} = newJBool true
    manifest{otherDir} = newJBool true

proc fileFunc (
  file: string, root: string, otherRoot: string, manifest: JsonNode
) =
  let otherFile = otherRoot / relativePath(file, root)
  let fileExists = fileExists file
  let otherFileExists = fileExists otherFile

  let fileIsDeleted = manifest{root}.getBool() and # A check to see if initialising
    ((manifest{file} != nil and manifest{file}.getBool() == false and not fileExists) or
    (manifest{otherFile} != nil and manifest{otherFile}.getBool() == false and not otherFileExists))

  if fileIsDeleted:
    if fileExists:
      echo fmt" - '{file}'"
    if otherFileExists:
      echo fmt" - '{otherFile}'"
    if not test:
      removeFile file
      removeFile otherFile
    manifest{file} = newJBool false
    manifest{otherFile} = newJBool false
  elif not existsFile otherFile:
    echo fmt" + '{otherFile}'"
    if not test:
      copyFile file, otherFile
    manifest{otherFile} = newJBool true
  else:
    let fileMTime = getLastModificationTime file
    let otherFileMTime = getLastModificationTime otherFile
    # Compare modified times:
    if fileMTime > otherFileMTime:
      echo fmt" + '{otherFile}'"
      if not test:
        copyFile file, otherFile
      manifest{otherFile} = newJBool true
    elif fileMTime < otherFileMTime:
      echo fmt" + '{file}'"
      if not test:
        copyFile otherFile, file
      manifest{file} = newJBool true

proc syncRepo (config: JsonNode, manifest: JsonNode): JsonNode =
  var manifest = manifest
  if manifest == nil:
    manifest = %* {}
  let
    start = cpuTime()
    ignore = parseRegex(config["ignore"].getStr())
    cloudDirPath = config["cloudDirPath"].getStr()
    localDirPath = config["localDirPath"].getStr()
    repoName = config["repoName"].getStr()
  
  echo fmt"Syncing {repoName}"

  discard existsOrCreateDir(cloudDirPath)
  discard existsOrCreateDir(localDirPath)

  let cloud = readdir(cloudDirPath, ignore = ignore)
  let local = readdir(localDirPath, ignore = ignore)

  updateManifest(manifest, ignore)

  for dir in cloud.dirs:
    directoryFunc dir, cloudDirPath, localDirPath, manifest
  for dir in local.dirs:
    directoryFunc dir, localDirPath, cloudDirPath, manifest

  for file in cloud.files:
    fileFunc file, cloudDirPath, localDirPath, manifest
  for file in local.files:
    fileFunc file, localDirPath, cloudDirPath, manifest

  updateManifest(manifest, ignore)
  manifest{cloudDirPath} = newJBool true
  manifest{localDirPath} = newJBool true

  echo fmt"{repoName}: {cpuTime() - start}"
  return manifest

proc syncLocalAndCloud (
  configPath: string,
  manifestPath: string,
  test: bool = false
): int =
  let start = cpuTime()

  sync.test = test
  if test:
    echo " * Running in TEST mode"

  var config = parseJson(readFile(configPath))
  var manifest = parseJson(readFile(manifestPath))
  # var r: openArray[FlowVarBase, string]

  parallel:
    for c in config.getElems():
      if not c{"active"}.getBool():
        continue
      let repoName = c{"repoName"}.getStr()
      # r.add(spawn syncRepo(c, manifest{repoName}))
      let partOfManifest = ^spawn syncRepo(c, manifest{repoName})
      manifest{repoName} = partOfManifest
  # Run serially:
  # for c in config.getElems():
  #   if not c{"active"}.getBool():
  #     continue
  #   let repoName = c{"repoName"}.getStr()
  #   manifest{repoName} = syncRepo(c, manifest{repoName})

  # echo blockUntilAny(r)
  # echo r

  echo "Finished syncing all repos!"

  for key, value in manifest.getFields().pairs:
    echo "key: ", key
    echo "value is nil: ", value == nil
    echo key, " ", value

  if not test:
    writeFile(manifestPath, manifest.pretty 2)
  else:
    writeFile("./test.json", manifest.pretty 2)
  echo "Saved manifest.json"

  echo fmt"sync.nim: {cpuTime() - start}"
  echo "Press ENTER to end..."
  discard readLine(stdin)

# Dispatch syncLocalAndCloud functon as a CLI:
dispatch(
  syncLocalAndCloud,
  help = {
    "configPath": "Path to the config",
    "manifestPath": "Path to the manifest",
    "test": "Test/debug mode"
  }
)
